src/metabase/query_processor/middleware/add_implicit_clauses.clj

```diff
@@ -117,14 +117,49 @@
 ;;; |                                        Add Implicit Breakout Order Bys                                         |
 ;;; +----------------------------------------------------------------------------------------------------------------+
 
+(defn- fix-order-by-field-refs
+  "This function transforms top level integer field refs in order by to corresponding string field refs from breakout
+  if present.
+
+  ## Context
+  In current situation, ie. model as a source, then aggregation and breakout, and finally order by a breakout field,
+  [[metabase.lib.order-by/orderable-columns]] returns field ref with integer id, while reference to same field, but
+  with string id is present in breakout. Then, [[add-implicit-breakout-order-by]] adds the string ref to order by.
+
+  Resulting query would contain both references, while integral is transformed differently -- it contains no casting.
+  As that is not part of group by, the query would fail.
+
+  Reference: https://github.com/metabase/metabase/issues/44653."
+  [{:keys [breakout order-by] :as inner-query}]
+  (if (or (empty? breakout) (empty? order-by))
+    inner-query
+    (let [name->breakout (into {}
+                               (keep (fn [[tag id-or-name :as clause]]
+                                       (when (and (= :field tag)
+                                                  (string? id-or-name))
+                                         [id-or-name clause])))
+                               breakout)]
+      (update inner-query :order-by
+              (partial reduce (fn [acc [dir [tag id-or-name] :as order-by]]
+                                (conj acc (or (and (= :field tag)
+                                                   (integer? id-or-name)
+                                                   (let [name ((some-fn :lib/desired-column-alias :name)
+                                                               (lib.metadata/field (qp.store/metadata-provider)
+                                                                                   id-or-name))]
+                                                     (when-some [breakout (name->breakout name)]
+                                                       [dir breakout])))
+                                              order-by)))
+                       [])))))
```

Comment:
This is hard to read. `(reduce (fn [acc x] (conj x ...)) [] xs)` is `mapv`, so let's start there.

In a case like this with nontrivial logic inside, I think `(update inner-query :order-by #(vec (for [order-by %] ...)))` would be easier to read, but factoring out the inner function is usually better.

The `(and ...)` case distilling the ref into the field name makes a nice factor. Then the `mapv` function is
```clojure
(if-let [breakout (-> inner-ref ref->field-name name->breakout))]
  [dir breakout]
  order-by)```

Reply:
Thanks for the pointer. I've modified the the code to use `mapv` and moved the inner logic into 2 functions bound in `let`. I haven't used `letfn` to avoid another nesting.